shell/client/notifications-client.js

```diff
@@ -0,0 +1,230 @@
+// Sandstorm - Personal Cloud Sandbox
+// Copyright (c) 2016 Sandstorm Development Group, Inc. and contributors
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+testNotifications = () => {
+  // Run on console to create some dummy notifications for the purpose of seeing what they look
+  // like.
+
+  Meteor.call("testNotifications");
+};
+
+const getNotificationPath = (notification) => {
+  if (notification.admin) {
+    return notification.admin.action;
+  } else if (notification.referral) {
+    return "/referrals";
+  } else if (notification.grainId) {
+    return "/grain/" + notification.grainId + (notification.path ? "/" + notification.path : "");
+  } else {
+    return null;
+  }
+};
+
+const removeTrailingSlash = (path) => {
+  while (path.slice(-1) === "/") {
+    path = path.slice(0, -1);
+  }
+
+  return path;
+};
+
+Tracker.autorun(function () {
+  // While the tab is visible, automatically dismiss any notifications that link to the current
+  // URL.
+
+  if (!browserTabHidden.get()) {
+    const path = removeTrailingSlash(currentPath.get());
+    Notifications.find().forEach((notification) => {
+      if (!notification.ongoing) {
+        const npath = getNotificationPath(notification);
+        if (npath) {
+          if (removeTrailingSlash(getNotificationPath(notification)) === path) {
+            Meteor.call("dismissNotification", notification._id);
+          }
+        }
+      }
+    });
+  }
+});
+
+Template.notificationsPopup.helpers({
+  notifications: function () {
+    Meteor.call("readAllNotifications");
```

Comment:
This call probably belongs in a `Template.notificationsPopup.onCreated(...)` block rather than in the middle of a template helper.